Execution interface cleanup

gerhardol opened the pull request:
Follow up to #9056, some discussions in #9476 too

## Proposed changes

* Execution: Adjust use of exit status, alloutput
Review that stderr (included in AllOutput) is not included in parsing of output data

* GetOutput only to be used where exit code is 0 and no error output

* GetOutputlines() removed, all callsites hanged on the call anyway.

Replace calls with throwOnErrorOutput:false with Execute()
Parse only StandardOutput for the replaced calls,
also where throwOnErrorOutput:false was set before.


## Test methodology <!-- How did you ensure quality? -->

manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -114,23 +110,21 @@ public static class ExecutableExtensions
                 redirectOutput: true,
                 outputEncoding,
-                throwOnErrorOutput: throwOnErrorOutput);
+                throwOnErrorOutput: true);
```
This line can be removed, too. Or do you want to emphasize it?

gerhardol replied:
That was the intention, Can remove if you find it confusing.
(I have another argument to remove, then it is only visible when false.)

mstv replied:
I have no preference.

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2152,8 +2155,7 @@ public async Task<IReadOnlyList<Remote>> GetRemotesAsync()
 
             static IReadOnlyList<Remote> ParseRemotes(ExecutionResult result)
             {
-                IEnumerable<string> lines = result.StandardOutput.LazySplit('\n', StringSplitOptions.RemoveEmptyEntries)
-                    .Concat(result.StandardError.LazySplit('\n', StringSplitOptions.RemoveEmptyEntries));
+                IEnumerable<string> lines = result.StandardOutput.LazySplit('\n', StringSplitOptions.RemoveEmptyEntries);
```
I guess the early exit on "not a git repository" below will not be triggered anymore. Should work anyway.

gerhardol replied:
Yes, those lines should be removed

mstv commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -1500,15 +1499,7 @@ public void ResetNoncommittedSelectedLines()
-            if (EnvUtils.RunningOnWindows())
-            {
-                // remove file mode warnings
-                Regex regEx = new("warning: .*has type .* expected .*", RegexOptions.Compiled);
```
❔ Have you removed `RegexOptions.Compiled` on purpose when moving it below?

gerhardol replied:
Yes, that should increase the startup time a lot more than decrease the runtime.
Most regex are static already, plan to look in to that later.

This was from #9468

mstv commented on GitUI/Script/ScriptRunner.cs:
```diff
@@ -134,7 +134,8 @@ private static CommandStatus RunScriptInternal(IWin32Window owner, IGitModule mo
                 {
-                    string revisionRef = new Executable(command, module.WorkingDir).GetOutputLines(argument).FirstOrDefault();
+                    ExecutionResult result = new Executable(command, module.WorkingDir).Execute(argument);
+                    string revisionRef = result.StandardOutput.Split('\n').FirstOrDefault();
```
```suggestion
                    string revisionRef = result.StandardOutput.LazySplit('\n').FirstOrDefault();
```

gerhardol replied:
Not available here

RussKie replied:
We'll need to see whether we always split it, and maybe move the split into `ExecuteAsync` changing `ExecutionResult` to something similar:
```cs
    public readonly struct ExecutionResult
    {
        public string[] StandardOutput { get; }
        public string[] StandardError { get; }
```

mstv reviewed (approved):
Looks sensible to me. Just nits / questions.

RussKie commented on Plugins/GitFlow/GitFlowForm.cs:
```diff
@@ -149,7 +149,7 @@ private IReadOnlyList<string> GetBranches(string typeBranch)
             var result = _gitUiCommands.GitModule.GitExecutable.Execute(args);
 
-            if (result.ExitCode != 0 || result.StandardOutput is null)
+            if (!result.ExitedSuccessfully || result.StandardOutput is null)
```
Can `result.StandardOutput is null` be ever true?

gerhardol replied:
The implementations of various `CleanString()` in `ExecutableExtensions` is not annotated to be null, so I guess that can me changed too.

RussKie commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -1557,12 +1548,20 @@ private void ApplySelectedLines(bool allFile, bool reverse)
+        private void ProcessApplyOutput(GitArgumentBuilder args, byte[] patch, bool patchUpdateDiff = false)
         {
+            ExecutionResult result = Module.GitExecutable.Execute(args, inputWriter => inputWriter.BaseStream.Write(patch), throwOnErrorOutput: false);
+            string output = result.AllOutput.Trim();
+            if (EnvUtils.RunningOnWindows())
+            {
+                // remove file mode warnings
+                Regex regEx = new("warning: .*has type .* expected .*");
+                output = output.RemoveLines(regEx.IsMatch);
+            }
```
I'm confused about this one... Warnings would be in the error stream, why not ignore those completely?

gerhardol replied:
AllOutput is stdout+stderr

RussKie replied:
That's my underlying question - is stderr used for anything useful? Previously the output would contain both of stdout and stderr, so we needed this filtering. Is it still required with this change?

gerhardol replied:
In this case, this is what we want to possibly display to the user.

See GitModule.createErrorGitItemStatus() where an "error item" is created from standard error at failures.

In #9468 stderr is used in popups at failures in the same way. In that PR I am not sure if this is too aggressive or not, so Trace will print a few more printouts (such printouts should be logged some way).

This "clean" could be lazy but the processing is a fraction of the time to patch files so it will not be noticeable.

RussKie reviewed (approved):
👍

gerhardol commented:
> We'll need to see whether we always split it, and maybe move the split into ExecuteAsync changing ExecutionResult to something similar:

Maybe 7 occurrences splits the output, a lot more occurrences (60?) is not.
Of course, this PR could provide a StandardOutputLines property in addition to StandardOutput.

gerhardol commented:
Rebased after merging of #9468, squashed the review comment.

https://github.com/gitextensions/gitextensions/pull/9488#discussion_r688944874 could be refactored
https://github.com/gitextensions/gitextensions/pull/9488#discussion_r688937769 could be changed too

gerhardol commented:
Merging later tonight

RussKie commented:
👍

On Thu, 19 Aug. 2021, 8:23 pm Gerhard Olsson, ***@***.***>
wrote:

> Merging later tonight
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9488#issuecomment-901795775>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXTFRZZHQXBTM6FKUVDT5TLSDANCNFSM5CB7JYMA>
> .
>
